crates/bindings-csharp/BSATN.Runtime/Builtins.cs

```diff
@@ -8,11 +8,136 @@ namespace SpacetimeDB;
 {
-    // Same as `Convert.ToHexString`, but that method is not available in .NET Standard
-    // which we need to target for Unity support.
-    public static string ToHex<T>(T val)
-        where T : struct =>
-        BitConverter.ToString(MemoryMarshal.AsBytes([val]).ToArray()).Replace("-", "");
+    /// <summary>
+    /// Convert this object to a BIG-ENDIAN hex string.
+    ///
+    /// Big endian is almost always the correct convention here. It puts the most significant bytes
+    /// of the number at the lowest indexes of the resulting string; assuming the string is printed
+    /// with low indexes to the left, this will result in the correct hex number being displayed.
+    ///
+    /// (This might be wrong if the string is printed after, say, a unicode right-to-left marker.
+    /// But, well, what can you do.)
+    ///
+    /// Similar to `Convert.ToHexString`, but that method is not available in .NET Standard
+    /// which we need to target for Unity support.
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    /// <param name="val"></param>
+    /// <returns></returns>
+    public static string ToHexBigEndian<T>(T val)
+        where T : struct => BitConverter.ToString(AsBytesBigEndian(val).ToArray()).Replace("-", "");
+
+    /// <summary>
+    /// Read a value of type T from the passed span, which is assumed to be in little-endian format.
+    /// The behavior of this method is independent of the endianness of the host, unlike MemoryMarshal.Read.
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    /// <param name="source"></param>
+    /// <returns></returns>
+    public static T ReadLittleEndian<T>(ReadOnlySpan<byte> source)
+        where T : struct => Read<T>(source, !BitConverter.IsLittleEndian);
+
+    /// <summary>
+    /// Read a value of type T from the passed span, which is assumed to be in big-endian format.
+    /// The behavior of this method is independent of the endianness of the host, unlike MemoryMarshal.Read.
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    /// <param name="source"></param>
+    /// <returns></returns>
+    public static T ReadBigEndian<T>(ReadOnlySpan<byte> source)
+        where T : struct => Read<T>(source, BitConverter.IsLittleEndian);
+
+    /// <summary>
+    /// Convert the passed byte array to a value of type T, optionally reversing it before performing the conversion.
+    /// If the input is not reversed, it is treated as having the native endianness of the host system.
+    /// (The endianness of the host system can be checked via System.BitConverter.IsLittleEndian.)
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    /// <param name="source"></param>
+    /// <param name="reverse"></param>
+    /// <returns></returns>
+    static T Read<T>(ReadOnlySpan<byte> source, bool reverse)
+        where T : struct
+    {
+        if (reverse)
+        {
+            Span<byte> reversed = stackalloc byte[source.Length];
```

Comment:
I'd suggest to instead: 1) read T, 2) create a byte view over it using [MemoryMarshal.CreateSpan](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.memorymarshal.createspan?view=net-8.0) + [MemoryMarshal.AsBytes](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.memorymarshal.asbytes?view=net-8.0), 3) reverse bytes in-place.

That way you can avoid intermediate array and copying.

Reply:
Good call.